### PR TITLE
DT tooling and DT API for RANGES property

### PR DESCRIPTION
**Pull request by superna9999**
As requested in https://github.com/zephyrproject-rtos/zephyr/pull/37668#pullrequestreview-729799792, this adds support for parsing and extracting the DT RANGES property with the associated macros.

The ranges property is used standalone in PCIe DT bindings, As described in IEEE Std 1275-1994, to describe the PCI I/O and memory regions.

### CODE REVIEW

**Comment by superna9999** [comment]
Seems doctree doesn't like #address-cells and #size-cells in doc, is there a way to do this correctly ?

**Comment by mbolivar-nordic** [comment]
> Seems doctree doesn't like #address-cells and #size-cells in doc, is there a way to do this correctly ?

You need to indent the code examples so doxygen recognizes them as preformatted code blocks. See the other headers for examples.

**Review by tbursztyka — APPROVED** [review]

**Comment by superna9999** [comment]
@galak is the API ok so I can add tests in order to merge #37688 ?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
temporary -1 until I've had a chance to review

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (mbolivar-nordic) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1172]: This seems to be special case code for PCIe, but there is no indication that the node is related to PCIe, so why is this here? I don't understand how it's safe to assume this in general.; nothing in the DT spec says that 3 is a magic value which means the first entry is always PCIe-specific flags, and this code runs for every node in the DT.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (mbolivar-nordic) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:1191]: I'm not familiar with names being used for the ranges property; can you please provide a reference to explain what's going on here?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Some initial comments on the edtlib.py changes. Thanks for your patience!

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1363]: yep, will fix

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1366]: Ranges in PCI is not obvious.

The PCI controller nodes uses the standard ranges property to declare memory ranges to be used with childs.

But these are not used directly, they are negociated with the BAR child pci registers, so translation is not to be done like the classic DT bus ranges.

Basically it uses the standard device tree ranges (as defined in https://buildmedia.readthedocs.org/media/pdf/devicetree-specification/latest/devicetree-specification.pdf pages 18/29) encoded as an arbitrary number of (`child-bus-address`
, `parent-bus-address`, `length`) triplets.

But the first `child-bus-address` has a different `#address-cells`, 2 or 3 depending on the parent node #address-cells`.
The extra cell encodes the memory ranges properties, this was defined a long time ago in IEEE Std 1275-1994 (https://www.openfirmware.info/data/docs/bus.pci.pdf page 4).

The main problem is that today the ranges cells are interpreted as C numbers in headers, with `#address-cells = 2` it's ok and the flags cell can be extracted in C, but with `#address-cells = 3` it's another issue.

I considered that `#address-cells = 3` is unique to PCI, and so splitting the extra cell into a separate `child_bus_flags` attribute was simpler.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1358]: ok

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1369]: ok

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1139]: ok will add

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1151]: ok

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1172]: (in ref to https://github.com/zephyrproject-rtos/zephyr/pull/38106#discussion_r718295619)
Yep I understand, but I don't see any other way to handle `#address-cells = 3`

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1191]: Isn't `add_names` necessary to be used from `gen_defines` ?

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1166]: My bad, Vim detected `range` as a reserved keyword... will remove.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1163]: ok

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1146]: ok

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1178]: >  Yep I understand, but I don't see any other way to handle `#address-cells = 3`

Is there some other way the node marks itself as a PCIe controller? Some sort of property or node name convention?

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1370]: Thanks for the reference; I will read through the IEEE doc.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1178]: ... and if not, perhaps we can get some extra hints from the bindings to use in edtlib. E.g. we can treat the node differently if and only if we see `bus: pci` (https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bus) in its matching binding.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1370]: For reference here is how u-boot handles the ranges attribute: https://github.com/u-boot/u-boot/blob/b5d7a200a81097fddef7ba5ec42e330d49fa2f33/drivers/pci/pci-uclass.c#L960

Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1370]: Here for Linux https://github.com/torvalds/linux/blob/02d5e016800d082058b3d3b7c3ede136cdc6ddcb/drivers/pci/of.c#L274

Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:1370]: The main difference it they handle the ranges node dynamically because they know the node is a pci node

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: @mbolivar-nordic OK I did that.

I only strictly parsed Ranges in edtlib (so address-cells generates 92bit numbers in python, but it's ok) and fixed the tests accordingly. I also passed the cells count for each part of a range, because it's used in the case of PCIe to know the width of the first range entry to extract the flags.

Then I simply used the bus bindings property in gen_defines to extract the flags.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py
Comment (superna9999) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: OK, I dropped this from edtlib, see https://github.com/zephyrproject-rtos/zephyr/pull/38106#discussion_r719232883

**Comment by superna9999** [comment]
Following the rework, I completely cleaned the `devicetree/ranges.h` header and the related tests.

**Comment by mbolivar-nordic** [comment]
Thanks @superna9999 . I'll get back to you this week; I need some time to digest.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I'm a fan of the latest edtlib patch. Thank you.

I took some time to try to understand the IEEE document as well and I think I have a rough idea now. Thanks again for the references.

Here are some more comments on the rest of the PR. I haven't looked at the test cases in detail yet, but this is what I'm thinking on the other unreviewed patches.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Looking good.

I found some typos and things in the docstrings, but in general I totally support the direction you took with where the macros go and how you are documenting them.

I will take a look at the tests tomorrow. After those are reviewed, I'll be ready to approve.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Test cases look great. Very thorough :).

I ended up having some style nitpicks and things that were easier to explain as a commit on top:

https://github.com/mbolivar-nordic/zephyr/commit/3b7e0a8341d1dd610f04783c1497b39d8fed98b4

Could you please take that patch? Then I think I'm ready to approve.

What do you think about the stable API change process? Did you want me to handle that for you, or did you want to do it yourself?

Here's a direct link to the steps in the process: https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#introducing-incompatible-changes

Almost there! :slightly_smiling_face:

**Comment by mbolivar-nordic** [comment]
> I'll try to start the stable API process, but, should I start it for this PR or should I open a separate PR for [9685ca8](https://github.com/zephyrproject-rtos/zephyr/commit/9685ca867640baa6a8380c87d5fc5212a3951950) ?
> 
> And do I need to also do the same process for [da0cfd6](https://github.com/zephyrproject-rtos/zephyr/commit/da0cfd618365f0d328a809714139595b33d23b61) ?

I think that in this case, it is OK to do a single stable API change process for both commits, since they are part of the same PR, which is making one logical change ("use DT_NUM_RANGES instead of DT_PROP_LEN for the ranges property") to the API.

**Review by mbolivar-nordic — APPROVED** [review]
Approving after marking DNM / Stable API Change / etc.

@superna9999, you just need to follow the process now. If nobody objects at the meeting where this is decided, you should be fine.

@nashif recently started discussion at the TSC about moving stable API changes to that forum from the API meeting, so I'm not sure what meeting you'll need to discuss it at, but that won't stop you from getting the process going.

**Review by jhedberg — APPROVED** [review]

**Review by galak — COMMENTED** [review]
What are we doing about dma-ranges?

**Comment by mbolivar-nordic** [comment]
> The minimal should be to exclude `dma-ranges` like `interrupt`, `reg` and `ranges` like in:

My opinion is that this is acceptable to include this while we're breaking things for regular `ranges`, and then defer `dma-ranges` support to potential future work.

If you do, you will need to update the stable API change request etc.

**Review by galak — COMMENTED** [review]
So let's add `dma-ranges` to the comment now so we don't need to do that in the future.  I don't think this requires going to the API meeting since nothing uses or supports `dma-ranges` so no issue.

**Comment by galak** [comment]
We should probably update release-notes as part of this PR.

**Comment by superna9999** [comment]
@galak @mbolivar-nordic done

**Comment by superna9999** [comment]
@mbolivar-nordic @galak Ping.

Are the `release-notes` and `dma-ranges` changes ok ?

**Comment by galak** [comment]
> @mbolivar-nordic @galak Ping.
> 
> Are the `release-notes` and `dma-ranges` changes ok ?

LGTM.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Looks like a merge conflict in release notes.

**Comment by superna9999** [comment]
@galak Fixed

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**
